src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1919,10 +1919,42 @@ multi_relation_restriction_hook(PlannerInfo *root, RelOptInfo *relOptInfo,
 	 * We're also keeping track of whether all participant
 	 * tables are reference tables.
 	 */
-	if (distributedTable)
+	if (isCitusTable)
 	{
 		cacheEntry = GetCitusTableCacheEntry(rte->relid);
 
+		/*
+		 * The statistics objects of the distributed table are not relevant
+		 * for the distributed planning, so we can override it.
+		 *
+		 * Normally, we should not need this. However, the combination of
+		 * Postgres commit 269b532aef55a579ae02a3e8e8df14101570dfd9 and
+		 * Citus function AdjustPartitioningForDistributedPlanning()
+		 * forces us to do this. The commit expects statistics objects
+		 * of partitions to have "inh" flag set properly. Whereas, the
+		 * function overrides "inh" flag. To avoid Postgres to throw error,
+		 * we override statlist such that Postgres does not try to process
+		 * any statistics objects during the standard_planner() on the
+		 * coordinator. In the end, we do not need the standard_planner()
+		 * on the coordinator to generate an optimized plan. We call
+		 * into standard_planner() for other purposes, such as generating the
+		 * relationRestrictionContext here.
+		 *
+		 * AdjustPartitioningForDistributedPlanning() is a hack that we use
+		 * to prevent Postgres' standard_planner() to expand all the partitions
+		 * for the distributed planning when a distributed partitioned table
+		 * is queried. It is required for both correctness and performance
+		 * reasons. Although we can eliminate the use of the function for
+		 * the correctness (e.g., make sure that rest of the planner can handle
+		 * partitions), it's performance implication is hard to avoid. Certain
+		 * planning logic of Citus (such as router or query pushdown) relies
+		 * heavily on the relationRestrictionList. If
+		 * AdjustPartitioningForDistributedPlanning() is removed, all the
+		 * partitions show up in the, causing high planning times for
+		 * such queries.
+		 */
+		relOptInfo->statlist = NIL;
```

Comment:
thanks for the thorough explanation

main danger is that we're in the middle of the planner, so if it relies on this being set based on earlier state then we're in trouble, but it doesn't seem like there are a lot of callers

Reply:
I can think of two options for the future:

- Do not use standard_planner() on the coordinator, only extract the restriction related information (as you noted in the private chat)
- Do not use AdjustPartitioningForDistributedPlanning(). I think this is doable, but we should be careful with the performance implications as Postgres generates restrict infos per partition. So `restrictionContext->relationRestrictionList` and `joinRestrictionContext->joinRestrictionList` becomes huge lists when # partitions is high

Reply:
Looking into this thread on psql-hackers (https://www.postgresql.org/message-id/flat/CAMbWs48pb_-fV%3DkG8CRVvZEHhCgTFUi6XY6UPK9Ob32LzH%2Bd8g%40mail.gmail.com#eec957e600172936e233aed44daf22af), I started to think whether our case can also be resolved with that patch.

Wanted to share to keep an eye on this